mentat/language_server.py

```diff
@@ -0,0 +1,253 @@
+
+    @property
+    def is_serving(self):
+        return self._server is not None and self._server.is_serving
+
+    @property
+    def is_stopping(self):
+        return self._stop_task is not None and not self._stop_task.done()
+
+    @property
+    def is_stopped(self):
+        return not self.is_serving and not self.is_stopping
+
+    async def _serve_tcp(self) -> None:
+        logger.info(f"Starting TCP server on {self.host}:{self.port}")
+        self._stop_event = threading.Event()
+        self._server = await self.loop.create_server(self.lsp, self.host, self.port)
+        try:
+            await self._server.serve_forever()
+        except asyncio.CancelledError:
+            pass
+        except (KeyboardInterrupt, SystemExit):
+            logger.debug("Got an exception")
+            pass
+        finally:
+            if self.is_serving and not self.is_stopping:
+                logger.debug("Killing server")
+                await self.stop()
+
+    def start(self) -> asyncio.Task[None]:
+        """Starts TCP server in a background asyncio.Task"""
+
+        def callback(_: asyncio.Task[None]):
+            self._server_task = None
+
+        self._server_task = self.loop.create_task(self._serve_tcp())
+        self._server_task.add_done_callback(callback)
+
+        return self._server_task
+
+    async def _stop(self) -> None:
+        if self._stop_event is not None:
+            self._stop_event.set()
+
+        if self._thread_pool:
+            # NOTE: might be blocking the async event loop
+            self._thread_pool.terminate()
+            self._thread_pool.join()
+
+        if self._thread_pool_executor:
+            # NOTE: might be blocking the async event loop
+            self._thread_pool_executor.shutdown()
+
+        if self._server:
+            self._server.close()
+            await self._server.wait_closed()
+            self._server = None
+
+    def stop(self) -> asyncio.Task[None]:
+        logger.info("Stopping the language server")
+
+        def callback(_: asyncio.Task[None]):
+            logger.info("The language server has stopped")
+            self._stop_task = None
+
+        self._stop_task = self.loop.create_task(self._stop())
+        self._stop_task.add_done_callback(callback)
+
+        return self._stop_task
+
+
+server = LanguageServer(host="127.0.0.1", port=7798, exit_on_lost_connection=False)
+
+
+@server.feature(lsp.INITIALIZE)
+async def initialize(ls: LanguageServer, params: lsp.InitializeParams):
+    if params.root_path is None:
+        logger.debug("No root path provided, using home directory")
+        ls.cwd = Path.home()
+    else:
+        ls.cwd = Path(params.root_path)
+
+
+@server.feature("mentat/createSession")
+async def create_session(ls: LanguageServer, params: lsp.InitializeParams):
+    assert ls.cwd is not None
+
+    ls.session = Session(cwd=ls.cwd)
+    ls.session.start()
+
+    async def handle_session_stream():
```

Comment:
Why do we need 2 separate listeners? Maybe I'm misunderstanding how this is supposed to work but I would assume that the point of this LanguageServer class is to forward stream messages to the vscode extension and forward responses from the vscode extension back to the session stream